### DISCUSSION TITLE
Improve chat interface

### DISCUSSION THREAD

**Discussion by Tostino** [bionic-gpt/bionic-gpt#229] [Ideas]
Hi there, I've been searching for a project similar to this one for a while! Very glad to have found it.

I was wondering if there were any plans on improving the actual chat interface? I've been checking out just about every UI I can. Something like [lobe-chat](https://github.com/lobehub/lobe-chat) seem to do a *pretty* good job at the actual chat interface, but they have none of the other features that I am looking for (the concept of users, auditing, etc).

The ability to edit / add / delete messages in the chain prior to generating the next response is a very useful feature IMO, but they don't do a good job of showing how the "conversation thread" evolved as the user made changes.

Also, there is the need for some (local) models to support roles other than system/user/assistant.
e.g. my fine tune has some extra roles available which hydrate specific variables within my prompt template. One of them is the current date, the other is the name of the task. I also have an extra "user_context" role that the model has been trained on using  for RAG type queries.

Here is an example curl call against a vLLM endpoint hosting my model:
```
curl http://0.0.0.0:8000/v1/chat/completions \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer " \
  -d '{
    "model": "Tostino/Inkbot-13B-8k-0.2",
    "stop": ["\n<#user#>", "\n<#bot#>"],
    "stream": false,
    "add_generation_prompt": true,
    "echo": false,
    "n": 1,
    "best_of": 1,
    "max_tokens": 500,
    "temperature": 0.5,
    "messages": [
	{"role": "meta-current_date", "content": "2023-10-20"},
	{"role": "meta-task_name", "content": "general"},
	{"role": "system", "content": "You are a helpful assistant. Please give very concise answers."},
	{"role": "user", "content": "What color is the car?"},
	{"role": "user_context", "content": "I did not get any sleep last night. My stupid neighbor was driving his stupid red corvette up and down the street, waking everyone up early in the morning."}
	]
  }'
  ```
  
I mainly want to put out feelers to see if you are open to contributions in this area? Have you put much thought into the final feature-set you would like to see for chat capability?

**Comment by ianpurton** [comment]
Hi Tostino,

> edit / add / delete messages

I like that idea we should definitely add that.

> Also, there is the need for some (local) models to support roles other than system/user/assistant.

I'm especially interested in this.

Can you show me your prompt template?

> I mainly want to put out feelers to see if you are open to contributions in this area?

We would take pull requests for everything you have said so far. We can also try and implement them ourselves but that will probably take longer.

> Have you put much thought into the final feature-set you would like to see for chat capability?

We're open to suggestions, what you have raised so far would give our users a great experience.

We haven't been concentrating on the chat interface lately as we have big piece of work for Open ID Connect that will go in soon and we've also been looking at getting Bioinic running on bare metal.

**Comment by Tostino** [comment]
Here is the chat template in a more readable format than stored in the tokenizer: https://github.com/vllm-project/vllm/blob/main/examples%2Ftemplate_inkbot.jinja


Understood on the openid connect support, I saw that quite large PR for it.

Also like the idea of being able to run on bare metal without k8s.

So, I'm not a rust dev myself, but I have hacked around in a couple of rust projects before. Enough to be dangerous, without actually knowing the language. So if I take a stab at any of these features, please forgive me if I'm just ignorant of rust conventions, or other parts of the ecosystem.

I do have experience though, I'm a pretty competent Java dev, know Postgres inside and out, and am pretty good at data modeling.


The other piece of the puzzle, is adding feedback support from the end users on the response (thumbs up / down, the category of issues they were seeing, etc). That data can then be used for RLHF if you so desire.  It's a missing part of the full lifecycle of an LLM for a tool like this.

**Comment by ianpurton** [comment]
> So, I'm not a rust dev myself,

No problem, [clippy](https://github.com/rust-lang/rust-clippy) will guide you.

> he other piece of the puzzle, is adding feedback support 

Agreed

**Comment by Tostino** [comment]
> No problem, [clippy](https://github.com/rust-lang/rust-clippy) will guide you.

Thanks, I will give that a shot!

I just got my development environment setup and devcontainers started. Going to make sure I can make changes and compile.

If you don't mind, I will spend a little bit of time building a mock up of my idea for what an ideal "power-user" chat interface, and editor would work, how they will interface with auditing, how it will interact with RLHF feedback and how we need to store that, etc. 

For example, I believe the editor should interact with the logits (or a useful visual representation of them), which would allow us to implement features such as: allow a user to pause generation, select some of the response, see alternate generations after a specific point in the response, and choose the correct one.
The official OpenAI API supports [0-5] logits on the response, but not on the initial prompt. Some inference servers support more features as far as logits are concerned, allowing you to get the logits on the users message as well as the response. 

So my idea is to have the chat component able to scale with the feature-set of the model/inference server, and be configurable to be as complex or simple as the user needs at the time (taking into account what is supported).
You should be able to specify all of the configurations on the model setup page, as well as other necessary configuration options. For example, stop tokens, which some models require to work properly in a chat context.
Another option would be to have configurable components setup in the "prompts" configuration which are able to configure the roles that the model supports, the UI components that they are tied to. E.g system, user, assistant...all pre-configured to the existing components. New roles could be configured. For an example: `user_context` could be tied to the standard editor window, or it could be tied to an upload component that accepts all types of text files. The `meta-task_name` role could be tied to a set of allowed values (Select component for the user to choose), or a hard coded value in the prompt configuration. `meta-current_date` could be tied to a date picker component for the user to set, or somehow we could make the current date/timestamp an available thing to set as a value.


Another feature that local inference servers can do, is allow you to complete an in-progress prompt from the user, by passing in the `add_generation_prompt=false` parameter and asking for a few tokens at a time. So this allows models that have a template that supports this feature (chatml doesn't by default, Inkbot does) to help the user complete their own prompt with an auto-complete type feature that can be hotkey activated, and possibly provide multiple options for them to choose from using the `n=2` argument for example.

I like a bit of what I've seen from other tools (intentionally forking the conversation to explore a different topic for example), but none have really incorporated a lot of what the API is capable of actually doing for you...they really just use it to create a response. There are many more uses while actually editing the text or composing your prompt.

Anyways, I was going to try and mock up some of the ideas so I can get some feedback before actually doing the work.

**Comment by ianpurton** [comment]
Sounds really great.

We do use [LiteLLM](https://litellm.ai/) in some scenarios as a proxy. They handle turning our chat into the prompt and negotiating with different inference engines.

Just so you are aware.

The production setup will use TGI as we need an engine that supports batching.

Hope this helps.
